Sources/Commands/SwiftTestTool.swift

```diff
@@ -36,111 +36,74 @@ extension TestError: CustomStringConvertible {
     }
 }
 
-public enum TestMode: Argument, Equatable, CustomStringConvertible {
-    case usage
-    case version
-    case listTests
-    case run(String?)
-    case runInParallel
-
-    public init?(argument: String, pop: @escaping () -> String?) throws {
-        switch argument {
-        case "--help", "-h":
-            self = .usage
-        case "-l", "--list-tests":
-            self = .listTests
-        case "-s", "--specifier":
-            guard let specifier = pop() else { throw OptionParserError.expectedAssociatedValue(argument) }
-            self = .run(specifier)
-        case "--version":
-            self = .version
-        case "--parallel":
-            self = .runInParallel
-        default:
-            return nil
+public class TestToolOptions: ToolOptions {
+    /// Returns the mode in with the tool command should run.
+    func mode() throws -> TestMode {
+        // If we got version option, just print the version and exit.
+        if printVersion {
+            return .version
+        }
+        // List the test cases.
+        if result.get(listTestOption) == true {
+            return .listTests
         }
+        // Run tests in parallel.
+        if result.get(parallelOption) == true {
+            return .runInParallel
+        }
+        return .run(result.get(specifierOption))
     }
+    private let listTestOption: OptionArgument<Bool>
+    private let parallelOption: OptionArgument<Bool>
+    private let specifierOption: OptionArgument<String>
 
-    public var description: String {
-        switch self {
-        case .usage:
-            return "--help"
-        case .listTests:
-            return "--list-tests"
-        case .run(let specifier):
-            return specifier ?? ""
-        case .version: return "--version"
-        case .runInParallel:
-            return "--parallel"
-        }
+    /// If the test target should be built before testing.
+    var buildTests: Bool {
+        return !(result.get(skipBuildOption) ?? false)
     }
-}
+    private let skipBuildOption: OptionArgument<Bool>
 
-public func ==(lhs: TestMode, rhs: TestMode) -> Bool {
-    return lhs.description == rhs.description
-}
+    override init(_ parser: ArgumentParser, args: [String]) throws {
```

Comment:
Yes, I thought we discussed this in the comments to a previous PR (I can't find it now though... perhaps that was on Slack?).  The idea would be to chain the parsers together.  And I don't think it's just a matter of flags that are common vs specific:  it seems to me that <cmd> <opt1> <subcmd> <opt2> is not interchangeable with <cmd> <subcmd> <opt1> <opt2>, i.e. each new subcommand start a new "scope".